calico/src/main/scala/calico/html/Codec.scala

```diff
@@ -64,48 +71,34 @@ object Codec {
-
-  object BooleanAsOnOffStringCodec extends Codec[Boolean, String] {
-    override def decode(domValue: String): Boolean = domValue == "on"
-    override def encode(scalaValue: Boolean): String = if scalaValue then "on" else "off"
-  }
-
-  object IterableAsSpaceSeparatedStringCodec extends Codec[Iterable[String], String] { // use for e.g. className
-    override def decode(domValue: String): Iterable[String] =
-      if domValue == "" then Nil else domValue.split(' ')
-    override def encode(scalaValue: Iterable[String]): String = scalaValue.mkString(" ")
-  }
-
-  object IterableAsCommaSeparatedStringCodec extends Codec[Iterable[String], String] { // use for lists of IDs
-    override def decode(domValue: String): Iterable[String] =
-      if domValue == "" then Nil else domValue.split(',')
-    override def encode(scalaValue: Iterable[String]): String = scalaValue.mkString(",")
-  }
-
-  object DoubleAsStringCodec extends Codec[Double, String] {
-    override def decode(domValue: String): Double =
-      domValue.toDouble // @TODO this can throw exception. How do we handle this?
-    override def encode(scalaValue: Double): String = scalaValue.toString
-  }
-
-  object IntAsStringCodec extends Codec[Int, String] {
-    override def decode(domValue: String): Int =
-      domValue.toInt // @TODO this can throw exception. How do we handle this?
-    override def encode(scalaValue: Int): String = scalaValue.toString
-  }
-
-}
+  val booleanAsAttrPresence: Codec[Boolean, String] = new:
+    def decode(domValue: String): Boolean = domValue ne null
+    def encode(scalaValue: Boolean): String = if scalaValue then "" else null
+
+  val booleanAsTrueFalseString: Codec[Boolean, String] = new:
+    def decode(domValue: String): Boolean = domValue == "true"
+    def encode(scalaValue: Boolean): String = if scalaValue then "true" else "false"
+
+  val booleanAsYesNoString: Codec[Boolean, String] = new:
+    def decode(domValue: String): Boolean = domValue == "yes"
+    def encode(scalaValue: Boolean): String = if scalaValue then "yes" else "no"
+
+  val booleanAsOnOffString: Codec[Boolean, String] = new:
+    def decode(domValue: String): Boolean = domValue == "on"
+    def encode(scalaValue: Boolean): String = if scalaValue then "on" else "off"
+
+  // hackery: Scala.js will transform `null` to `0` unless we erase the type
+
+  inline def doubleAsString: Codec[Double, String] =
+    doubleAsStringImpl.asInstanceOf[Codec[Double, String]]
+  private val doubleAsStringImpl: Codec[Any, String] = new:
+    def decode(domValue: String): Any =
+      domValue.toDoubleOption.getOrElse(null)
+    def encode(scalaValue: Any): String = scalaValue.toString
+
+  inline def intAsString: Codec[Int, String] =
+    intAsStringImpl.asInstanceOf[Codec[Int, String]]
+  private val intAsStringImpl: Codec[Any, String] = new:
+    def decode(domValue: String): Any =
+      domValue.toIntOption.getOrElse(null)
+    def encode(scalaValue: Any): String = scalaValue.toString
```

Comment:
This hack works as long as your are in generic types. As soon as you know that the type is `Int` or `Double`, then Scala.js will do the `null`-to-`0` conversion.

But maybe this is okay: as long as we do all the error handling on a generic codec (without knowing the precise type) then it will work okay in practice. It's a bit fickle.